### DISCUSSION TITLE
Add ability to set window_bits when using rust backend?

### DISCUSSION THREAD

**Discussion by praveenperera** [rust-lang/flate2-rs#410] [Q&A]
I was wondering if it would be possible to support `new_with_window_bits` function without needing the `zlib` or `zlib-ng` features.

Seems like the option for setting window_bits is available in `miniz_oxide`

**Comment by Byron** [comment]
You can probably [wire it up](https://docs.rs/flate2/latest/flate2/struct.Compress.html#method.new_with_window_bits) and see if [miniz_oxide](https://docs.rs/miniz_oxide/latest/miniz_oxide/enum.DataFormat.html#method.from_window_bits) can fully represent it.

Probably there are tests for this already, but it would be good to double-check that `miniz_oxide` is truly producing the expected output when window bits are set, just to be sure there is no actual reason for this omission. And assuming there is none, the result of such research should be a great PR.

**Reply by praveenperera** [reply]
Is this why `miniz_oxide` is not used with custom window bits? https://github.com/Frommi/miniz_oxide/blob/0a33effd414711b379e01b0613ba5ae85a0e14d0/src/lib_oxide.rs#L144-L148

**Comment by oyvindln** [comment]
You could add it though just like the original miniz the window bits argument doesn't do anything other than make the stream use a deflate wrapper if window bits is positive or no wrapper if it's negative (which the window_bits function in flate2 doesn't let you do directly.) If you for some reason want to use new_with_window_bits instead of the normal init function and init it with the standard window size I guess you could just add it and not that there is no reason to use it and the window size argument does not have any effect with the miniz_oxide backend (though I have no idea why you would want that.)

For it to be useful you would need implement support in miniz_oxide for compressing with smaller than the max window size (and decompressing with smaller buffers and/or erroring on encountering the wanted max window size) though that is an very niche use case. It was mainly useful back in the day in very memory limited situations and if you are in that sort of realm you probably would want to use miniz_oxide or zlib in no_std mode directly.

If there is a need for it feel free to contribute to miniz_oxide, I'm not very active in contributing myself at the moment.

**Reply by praveenperera** [reply]
I'm implementing this spec in rust: https://github.com/coinkite/BBQr/blob/master/BBQr.md#advanced-encodings

> You must use [zlib](https://www.zlib.net/) and provide a wbits value of 10. No Gzip nor Zlib file header should be included, and they are not needed since the wbits value is fixed. The compression level should typically be set to maximum compression effort (9) but the fixed wbits value limits this somewhat. We restrict wbits to this value because it defines the amount of memory the decoder will need to decompress the data.

So I'm assuming in this case, I would need to set the wbits manually?
